### PR TITLE
ENH: Mutliple Equations/MNLogit wrapping improvements

### LINKED ISSUE #2147: Use better equation names in MNLogit results
Now that we're treating factor Y inputs, we need to use the equation names for the columns of the results parameters.

(**Edit**: I think the above is misleading. AFAICT, we don't currently, and maybe never did, allow factors for Y. This is related to #2049 and #1210. I abandoned the PR #2197 because it introduced too much complexity. Could be worth revisiting at some point.)

### PR DESCRIPTION
**Pull request by jseabold**
Closes #2147.

### CODE REVIEW

**Comment by jseabold** [comment]
As part of this PR, I'd like to add support for selection of a reference category for MNLogit and mention the reference in the summary. Should be easy to do. #837

Needs to be added to release notes as well.

**Comment by josef-pkt** [comment]
@jseabold 
I'd like to move this to 0.8 
still several failures, changes or affects code all over the place, option for reference category is not working yet correctly
(and I'm not able to read through it very fast)

File: statsmodels/discrete/discrete_model.py [hunk 1]
```diff
@@ -502,22 +559,66 @@ def _derivative_exog(self, params, exog=None, transform='dydx',
 class MultinomialModel(BinaryModel):
 
     def _handle_data(self, endog, exog, missing, hasconst, **kwargs):
+        reference = kwargs.pop('reference', 0)
```
Comment (jbrockmendel) [inline statsmodels/discrete/discrete_model.py]: Can we just add `reference=0` to the signature?  Pushing back against the `**kwargs` pattern.

File: statsmodels/discrete/discrete_model.py [hunk 2]
```diff
@@ -502,22 +559,66 @@ def _derivative_exog(self, params, exog=None, transform='dydx',
...
+            if not isinstance(reference, int):
+                reference_idx = _get_index_from_list(ynames, reference)
+            else:
+                reference_idx = reference
+                reference = ynames[reference]
             yname = 'y'
+            ynames = ["{0}={1}".format(yname, ynames[i]) for i in keys]
         elif data_tools._is_using_pandas(endog, None):
+            # patsy goes through here
             endog_dummies, ynames, yname = _pandas_to_dummies(endog)
```
Comment (jbrockmendel) [inline statsmodels/discrete/discrete_model.py]: Not necessarily in this PR, but for pd.Categoricals a corner case to watch out for is when there are unused categories.

File: statsmodels/discrete/discrete_model.py [hunk 3]
```diff
@@ -2418,13 +2533,6 @@ def aic(self):
     def bic(self):
         return -2*self.llf + np.log(self.nobs)*(self.df_model+1)
 
-    def _get_endog_name(self, yname, yname_list):
```
Comment (jbrockmendel) [inline statsmodels/discrete/discrete_model.py]: +1 for making this not-a-method

File: statsmodels/base/data.py [hunk 4]
```diff
@@ -365,41 +365,38 @@ def _check_integrity(self):
...
+            names = how[1:]
         else:
+            names = ()
+            method_name = "attach_{0}".format(how)
+
+        if not hasattr(self, method_name):
             return obj
 
-    def attach_columns(self, result):
+        return getattr(self, method_name)(obj, *names)
```
Comment (jbrockmendel) [inline statsmodels/base/data.py]: -1. There's too much `getattr` and `setattr` as it is.

File: statsmodels/base/data.py [hunk 5]
```diff
@@ -365,41 +365,38 @@ def _check_integrity(self):
...
+
+    def attach_columns(self, result, *arg, **kwargs):
```
Comment (jbrockmendel) [inline statsmodels/base/data.py]: Are there existing cases when `**kwargs` get passed?

File: statsmodels/discrete/discrete_model.py [hunk 6]
```diff
@@ -1642,7 +1744,7 @@ class MNLogit(MultinomialModel):
     Notes
     -----
     See developer notes for further information on `MNLogit` internals.
-    """ % {'extra_params' : base._missing_param_doc}
+    """ % {'extra_params' : _mnlogit_extra_params}
```
Comment (jbrockmendel) [inline statsmodels/discrete/discrete_model.py]: If this is only used in one place, does why not just put it here?

**Comment by josef-pkt** [comment]
This pull request **introduces 2 alerts** and **fixes 1** when merging 778a7a4d92fff1deac3fc70ab8df1b0697b8d747 into f4cf57c1d757ef06baefc480a7cdad93ba617089 - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-29a54dfac403fe737952ab608c535a3b6d7897c7)

**new alerts:**

* 2 for Module\-level cyclic import

**fixed alerts:**

* 1 for Module\-level cyclic import

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Pull request closed without merging**